draft-ietf-mls-protocol.md

```diff
@@ -1524,223 +1460,294 @@ The creator of the group constructs an Init message as follows:
+actually removed before any application information is transmitted.
+
+The sender of a Commit message MUST include in it all valid Proposals that the
+sender has received during the current epoch.  Invalid Proposals include, for
+example, Proposals with an invalid signature or Proposals that are semantically
+inconsistent, such as a Remove proposal for an unoccupied leaf. The Commit MUST
+NOT combine Proposals sent within different epochs.  Despite these requirements,
+it is still possible for a valid Proposal not to be covered by a Commit, e.g.,
+because the sender of the Commit did not receive the Proposal.  In such cases,
+the sender of the proposal can retransmit the Proposal in the new epoch.
+
+Each proposal covered by the Commit is identified by a ProposalID structure.
+The `sender` field in this structure indicates the member of the group that sent
+the proposal (according to their index in the ratchet tree).  The `hash` field
+contains the first four octets of the hash of the Proposal structure itself,
+using the hash function for the group's ciphersuite.
 
 ~~~~~
 struct {
+    uint32 sender;
+    opaque hash[4];
+} ProposalID;
+
+struct {
+    ProposalID updates<0..2^16-1>;
+    ProposalID removes<0..2^16-1>;
+    ProposalID adds<0..2^16-1>;
     DirectPath path;
-} Update;
+} Commit;
 ~~~~~
 
-The sender of an Update message creates it in the following way:
+[[ OPEN ISSUE: This structure loses the welcome_info_hash, because new
+participants are no longer expected to have access to the Commit message adding
+them to the group.  It might be we need to re-introduce this assumption, though
+it seems like the information confirmed by the welcome_info_hash is confirmed at
+the next epoch change anyway. ]]
 
-* Generate a fresh leaf key pair
-* Compute its direct path in the current ratchet tree
+A member of the group applies a Commit message by taking the following steps:
 
-A member receiving a Update message first verifies
-the signature on the message, then updates its state as follows:
+1. Verify that the `epoch` field of the enclosing MLSPlaintext message is equal
+   to the `epoch` field of the current GroupContext object
 
-* Update the cached ratchet tree by replacing nodes in the direct
-  path from the updated leaf using the information contained in the
-  Update message
+2. Verify that the signature on the MLSPlaintext message verifies using the
+   public key from the credential stored at the leaf in the tree indicated by
+   the `sender` field.
+
+3. Generate a provisional GroupContext object by applying the proposals
+   referenced in the commit object in the order provided, as described in
+   {{proposals}}.  Add proposals are applied left to right: Each Add proposal is
+   applied at the leftmost unoccupied leaf, or appended to the right edge of the
+   tree if all leaves are occupied.
+
+4. Process the `update_path` value to update the ratchet tree referenced by the
```

Comment:
Typo: `path`

Also, if there's only a single DirectPath per epoch, won't this lead to trees becoming sparser (i.e., having fewer filled nodes) over time? It's fine as long as the root node is filled, but DirectPath messages might have to get bigger to accommodate the sparseness.

Reply:
The sparseness of the tree depends on how often all the members Commit. For example, if you have a group where only one person ever Commits, then the direct path above that person is the only one that will ever be populated, and MLS will degrade to "encrypt every message to every person individually" scheme.

If you have an application-level understanding of who needs to Commit and who's allowed to be lazy, it could be fine(?). It gives you a performance trade-off.